python/descriptor.c

```diff
@@ -38,113 +42,919 @@
+  PyObject* range_list = PyList_New(n);
+
+  for (int i = 0; i < n; i++) {
+    const upb_extrange* range = upb_msgdef_extrange(self->def, i);
+    PyObject* start = PyLong_FromLong(upb_extrange_start(range));
+    PyObject* end = PyLong_FromLong(upb_extrange_end(range));
+    PyList_SetItem(range_list, i, PyTuple_Pack(2, start, end));
+  }
+
+  return range_list;
+}
+
+static PyObject* PyUpb_Descriptor_GetExtensions(PyObject* _self,
+                                                void* closure) {
+  PyUpb_DescriptorBase* self = (void*)_self;
+  static PyUpb_GenericSequence_Funcs funcs = {
+      (void*)&upb_msgdef_nestedextcount,
+      (void*)&upb_msgdef_nestedext,
+      (void*)&PyUpb_FieldDescriptor_Get,
+  };
+  return PyUpb_GenericSequence_New(&funcs, self->def, self->pool);
+}
+
+static PyObject* PyUpb_Descriptor_GetExtensionsByName(PyObject* _self,
+                                                      void* closure) {
+  PyUpb_DescriptorBase* self = (void*)_self;
+  static PyUpb_ByNameMap_Funcs funcs = {
+      {
+          (void*)&upb_msgdef_nestedextcount,
+          (void*)&upb_msgdef_nestedext,
+          (void*)&PyUpb_FieldDescriptor_Get,
+      },
+      (void*)&PyUpb_Descriptor_LookupNestedExtension,
+      (void*)&upb_fielddef_name,
+  };
+  return PyUpb_ByNameMap_New(&funcs, self->def, self->pool);
+}
+
+static PyObject* PyUpb_Descriptor_GetEnumTypes(PyObject* _self, void* closure) {
+  PyUpb_DescriptorBase* self = (void*)_self;
+  static PyUpb_GenericSequence_Funcs funcs = {
+      (void*)&upb_msgdef_nestedenumcount,
+      (void*)&upb_msgdef_nestedenum,
+      (void*)&PyUpb_EnumDescriptor_Get,
+  };
+  return PyUpb_GenericSequence_New(&funcs, self->def, self->pool);
+}
+
+static PyObject* PyUpb_Descriptor_GetOneofs(PyObject* _self, void* closure) {
+  PyUpb_DescriptorBase* self = (void*)_self;
+  static PyUpb_GenericSequence_Funcs funcs = {
+      (void*)&upb_msgdef_oneofcount,
+      (void*)&upb_msgdef_oneof,
+      (void*)&PyUpb_OneofDescriptor_Get,
+  };
+  return PyUpb_GenericSequence_New(&funcs, self->def, self->pool);
+}
+
+static PyObject* PyUpb_Descriptor_GetOptions(PyObject* _self, PyObject* args) {
+  PyUpb_DescriptorBase* self = (void*)_self;
+  return PyUpb_DescriptorBase_GetOptions(
+      self, upb_msgdef_options(self->def),
+      &google_protobuf_MessageOptions_msginit,
+      "google.protobuf.MessageOptions");
+}
+
+static PyObject* PyUpb_Descriptor_CopyToProto(PyObject* _self,
+                                              PyObject* py_proto) {
+  return PyUpb_DescriptorBase_CopyToProto(
+      _self, (PyUpb_ToProto_Func*)&upb_MessageDef_ToProto,
+      &google_protobuf_DescriptorProto_msginit, py_proto);
+}
+
+static PyObject* PyUpb_Descriptor_EnumValueName(PyObject* _self,
+                                                PyObject* args) {
+  PyUpb_DescriptorBase* self = (void*)_self;
+  const char* enum_name;
+  int number;
+  if (!PyArg_ParseTuple(args, "si", &enum_name, &number)) return NULL;
```

Comment:
should you call some kind of PyErr on failure here?

Reply:
No, PyArg_ParseTuple() handles this.